### ISSUE TITLE
[UB/Underspec] What do the Stage Flags need to be for the "other half" of a pipeline barrier which does an ownership transfer?

### ISSUE THREAD

**Issue by devshgraphicsprogramming** [KhronosGroup/Vulkan-Docs#2319]
By the "other half" I mean the `dst` fields on the Release barrier, and `src` fields on the Acquire barrier.

The spec tells us the access masks should be 0, but doesn't tell us anything about the pipeline masks:
- are they meant to represent the stages of commands immediately before/after the acquire/release operation in queue submission order on the queue we'll submit them to?
- or are they meant to represent the stages on the other side of the release/acquire in the queue submission that will acquire/release on the other queue family?

AFAIK in the old spec, there used to be a requirement that the barriers executed match 100% (including stage and access flags) between the acquiring and releasing queue, which makes me think that the intended semantics was the second option.

IMHO the first option doesn't make sense, because it should be a given that we can't have accesses to a resource after-a-release or before-an-acquire without another pipeline barrier inbetween that re-acquires the resource. Unless you mean an implicit acquire by just using the resource, so effectively resulting in a possible content-wipe. I can see some corner-case strange setups like implicit acquire on Queue A, implicit acquire on Queue B but explicit transfer to Queue A afterwards, in this way you'd need the formal-acquire in a pipeline barrier to happen before the implicit acquire.

~~However the second option makes no sense, as ordering between Acquire/Release must be ensured by signalling and waiting a semaphore between the two submits on different queues that will release and acquire ownership. The semaphore wait/signal give you the `happens-before` relationship that a pipeline barrier would, so any execution barriers with stage masks from both queue commands seem redundant.~~ Forgot that semaphores have flags in the signal/wait infos on submits, but when would the flags in the semaphore signal/wait pair across the queue families NOT contain the stage flags in the two release/acquire barriers?

The first option ~~also~~ kind-of makes no sense, because I don't see a way to acquire or release the same resource range twice without a race condition. Why? Because to acquire twice, someone would have to release between the acquires, and the only way to ensure the ordering is with a semaphore, so this means the two acquires have to be in separate submits.

TL;DR  in my opinion, the src/dst pipeline masks should be 0 for an acquire/release.

This is what we're doing:
- acquire
```json
{
  "index": 267,
  "function": {
    "name": "vkCmdPipelineBarrier2",
    "thread": 1,
    "cmd_index": 3,
    "args": {
      "commandBuffer": 29,
      "pDependencyInfo": {
        "sType": "VK_STRUCTURE_TYPE_DEPENDENCY_INFO",
        "dependencyFlags": "0x00000000",
        "memoryBarrierCount": 0,
        "pMemoryBarriers": null,
        "bufferMemoryBarrierCount": 0,
        "pBufferMemoryBarriers": null,
        "imageMemoryBarrierCount": 1,
        "pImageMemoryBarriers": [
          {
            "sType": "VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER_2",
            "srcStageMask": "0x0000000100000000",
            "srcAccessMask": "0x0000000000001000",
            "dstStageMask": "0x0000000000000000", // is this wrong!?
            "dstAccessMask": "0x0000000000000000",
            "oldLayout": "VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL",
            "newLayout": "VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL",
            "srcQueueFamilyIndex": 1,
            "dstQueueFamilyIndex": 2,
            "image": 14,
            "subresourceRange": {
              "aspectMask": "0x00000001",
              "baseMipLevel": 0,
              "levelCount": 1,
              "baseArrayLayer": 0,
              "layerCount": 1
            },
            "pNext": null
          }
        ],
        "pNext": null
      }
    }
  }
},
```
- release 
```json
{
  "index": 278,
  "function": {
    "name": "vkCmdPipelineBarrier2",
    "thread": 1,
    "cmd_index": 2,
    "args": {
      "commandBuffer": 39,
      "pDependencyInfo": {
        "sType": "VK_STRUCTURE_TYPE_DEPENDENCY_INFO",
        "dependencyFlags": "0x00000000",
        "memoryBarrierCount": 0,
        "pMemoryBarriers": null,
        "bufferMemoryBarrierCount": 0,
        "pBufferMemoryBarriers": null,
        "imageMemoryBarrierCount": 2,
        "pImageMemoryBarriers": [
          {
            "sType": "VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER_2",
            "srcStageMask": "0x0000000000000000",  // is this wrong!?
            "srcAccessMask": "0x0000000000000000",
            "dstStageMask": "0x0000000000000800",
            "dstAccessMask": "0x0000000100000000",
            "oldLayout": "VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL",
            "newLayout": "VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL",
            "srcQueueFamilyIndex": 1,
            "dstQueueFamilyIndex": 2,
            "image": 14,
            "subresourceRange": {
              "aspectMask": "0x00000001",
              "baseMipLevel": 0,
              "levelCount": 1,
              "baseArrayLayer": 0,
              "layerCount": 1
            },
            "pNext": null
          },
```

Full Graphics Reconstruct on an Nvidia 1050 (needs to be replayed with validation):
[gfxrecon_capture_20240305T172423.zip](https://github.com/KhronosGroup/Vulkan-Docs/files/14498586/gfxrecon_capture_20240305T172423.zip)

This will produce a validation bug which IMHO is a false positive https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/7642 because there's only supposed to be one layout transition done between the Release and Acquire, the spec even (!) spells it out for us.

**Comment by krOoze** [comment]
It is a barrier as any other, and the stages need to be whatever it is to ensure serialization of those two barriers (typically via semaphore).

The memory dependency (and layout transition) is special. The execution dependency is normal. Keep in mine the stage flags are not even in the same `struct` for `vkCmdPipelineBarrier`, and it could even easily have items that are not transfering ownership.

**Comment by devshgraphicsprogramming** [comment]
@krOoze my question is about pipeline stage masks not access masks, I know the latter should be 0.

```
SrcStages make SrcAccesses Available -> QFOT Release -> No visibility needed as no one can access without UB, but What Stages!?
No writes to make available, but What Stages!? -> QFOT Acquire -> Visible to DstAccesses of Dst Stages
```

The semantics for me are completely unclear.

After quite some digging I found that the problem is a fundamental one.

Layout Transitions and Queue Family Ownership Transfers don't have a stage, so its a real head-scratcher what the masks should be.

Especially for the QFOT because as I've laid out, unlike a Layout Transition, it makes no sense to synchronize it with ANY other command in the same submit, the only thing they should synchronize with is the semaphore signal/wait after/before an ownership release/acquire.

I find the biggest shortcoming of the Vulkan Spec is that there exist non-action commands like Layout Transitions and Queue Family Ownership Transfers that do perform memory operations but they have no stage. Also semaphore signal/wait didn't use to have a stage, and now they kind-of do with sync2.

After a lot of asking and thinking, it seems like the "other half" stage masks for the QFOT can be always NONE, but the Semaphore Stage should be ALL_COMMANDS_BIT because of the aformentioned issue that a QFOT doesn't happen in a stage.

**Comment by devshgraphicsprogramming** [comment]
I guess the whole issue is just #347 but re-heated with sync2 barriers and sync2 semaphore signal/wait stage masks.

**Comment by HansKristian-Work** [comment]
To summarize how it works:

When releasing ownership, stage should be BOTTOM_OF_PIPE (or NONE in sync2). After releasing ownership it's not going to be used, and dstStageMask = BOTTOM_OF_PIPE does not block any work, so it is the ideal thing to do. Signaling a semaphore with ALL_COMMANDS_BIT waits for the release to complete. In sync1, signaling a semaphore implies ALL_COMMANDS_BIT. If using a specific stage like COLOR_ATTACHMENT_OUTPUT_BIT for the signal, then the release operation should use dstStage = COLOR for example.

For acquire, srcStage needs to be tied to the wait semaphore's pDstWaitStages for the dependencies to link up.

**Comment by devshgraphicsprogramming** [comment]
> When releasing ownership, dstStage should be BOTTOM_OF_PIPE (or NONE in sync2). After releasing ownership it's not going to be used, and dstStageMask = BOTTOM_OF_PIPE does not block any work, so it is the ideal thing to do. Signaling a semaphore with ALL_COMMANDS_BIT waits for the release to complete. While the release operation itself is performed in no particular stage, we observe that it is complete after waiting for dstStageMask. All image memory barriers behave similarly.

yeah this I managed to deduce, after many hours of head scratching and spec lawyering.

Would be nice if the spec mentioned this explicitly rather than "leaving it as an excercise to the reader".

**Comment by krOoze** [comment]
Omissions in a specification are as important as what it includes. A specification is not a tutorial. If it doesn't say the stagemask is restricted, then it is not restricted. If it doesn't say the stages behave in some special way, then they behave in the normal\ordinary way.

**Comment by HansKristian-Work** [comment]
I think a NOTE mentioning that dstStage = BOTTOM/NONE is expected and normal is fine. @krOoze were you going to propose a PR? If not I'll do it.

**Comment by krOoze** [comment]
@HansKristian-Work I mean, virtually any stage is fine. Stages have nothing to do with QFOT at all. Only relation they have to QFOT is that specification mandates the app to synchronize the two QFOT ops any which way the app desires (which potentially implies working with Stages).

I guess it is partly fault of sync2 that in the name of convenience it made things quite a bit more confusing. In sync1 stages were never part of a memory barrier. Only confusion there is there was a typo in the spec like 5 years ago referring accidentally to stages instead of accesses wrt QFOT.

IMO all information is there in the chapter. I was thinking more along better organizing the information. As it is, it seems the chapter jumps from concept to concept often interrupted by other paragraph or sentence. If you wish to actually add something to the chapter, then go ahead.

**Comment by oddhack** [comment]
This should be fixed in the 1.3.283 spec update.

**Comment by Trider12** [comment]
For future readers - as of now the spec clearly says that the `dstStageMask`/`srcStageMask` i.e the second/first synchronization scopes of the release and acquire barriers are ignored and that an `ALL_COMMANDS` semaphore **must** be used to synchronize them.

From Vulkan spec 1.3.287:
7.7.4. Queue Family Ownership Transfer:
Release:
> dstAccessMask is ignored for such a barrier, such that no visibility operation is executed - the value of this mask does not affect the validity of the barrier. The release operation happens-after the availability operation. **dstStageMask is also ignored for such a barrier** as defined by [buffer memory ownership transfer](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#buffer-memory-barrier-ownership-transfer) and [image memory ownership transfer](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#image-memory-barrier-ownership-transfer).

Acquire:
> srcAccessMask is ignored for such a barrier, such that no availability operation is executed - the value of this mask does not affect the validity of the barrier. The acquire operation happens-before the visibility operation. **srcStageMask is also ignored for such a barrier** as defined by [buffer memory ownership transfer](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#buffer-memory-barrier-ownership-transfer) and [image memory ownership transfer](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#image-memory-barrier-ownership-transfer). As the first synchronization scope for an acquire operation is empty there is no happens-before dependency. Such a dependency can be introduced by using VK_PIPELINE_STAGE_ALL_COMMANDS_BIT.

Note:
> Since a release and acquire operation does not synchronize with second and first scopes respectively, the VK_PIPELINE_STAGE_ALL_COMMANDS_BIT stage must be used to wait for a release operation to complete. Typically, a release and acquire pair is performed by a [VkSemaphore](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkSemaphore) signal and wait in their respective queues. Signaling a semaphore with [vkQueueSubmit](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vkQueueSubmit) waits for VK_PIPELINE_STAGE_ALL_COMMANDS_BIT. With [vkQueueSubmit2](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vkQueueSubmit2), stageMask for the signal semaphore must be VK_PIPELINE_STAGE_ALL_COMMANDS_BIT. Similarly, for the acquire operation, waiting for a semaphore must use VK_PIPELINE_STAGE_ALL_COMMANDS_BIT to make sure the acquire operation is synchronized.
